File: src/lib/crypto.ts
/**
 * Encryption and security management for ConfigSync.
 *
 * Uses AES-256-GCM (replacing the legacy Fernet/AES-128-CBC from the Python version).
 * Wire format for encrypted buffers: [IV 12 bytes][authTag 16 bytes][ciphertext].
 */

import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12;
const AUTH_TAG_LENGTH = 16;
const SALT_LENGTH = 32;
const KEY_LENGTH = 32;
const PBKDF2_ITERATIONS = 100_000;
const PBKDF2_DIGEST = 'sha256';
const VERIFICATION_STRING = 'CONFIGSYNC_VERIFICATION';

export default class CryptoManager {
  private configDir: string;
  private keyFile: string;
  private saltFile: string;
  private derivedKey: Buffer | null = null;

  constructor(configDir: string) {
    this.configDir = configDir;
    this.keyFile = path.join(configDir, '.key');
    this.saltFile = path.join(configDir, '.salt');
  }

  // ---------------------------------------------------------------------------
  // Lifecycle
  // ---------------------------------------------------------------------------

  /**
   * Initialize the crypto system with a master password.
   * Generates a random salt, derives a key, and stores a verification token.
   */
  initialize(password: string): void {
    if (fs.existsSync(this.keyFile)) {
      throw new Error("Crypto already initialized. Use 'unlock' instead.");
    }

    // Generate a random salt and persist it
    const salt = crypto.randomBytes(SALT_LENGTH);
    fs.writeFileSync(this.saltFile, salt);
    fs.chmodSync(this.saltFile, 0o600);

    // Derive the master key
    const key = this.deriveKey(password, salt);

    // Encrypt a known verification string so we can validate the password later
    const verification = this.encryptWithKey(Buffer.from(VERIFICATION_STRING, 'utf-8'), key);

    const keyData = {
      verification: verification.toString('base64'),
      version: '2.0', // v2 = AES-256-GCM
    };

    fs.writeFileSync(this.keyFile, JSON.stringify(keyData));
    fs.chmodSync(this.keyFile, 0o600);

    this.derivedKey = key;
  }

  /**
   * Unlock the crypto system by verifying the master password.
   */
  unlock(password: string): void {
    if (this.derivedKey !== null) {
      return; // already unlocked
    }

    if (!fs.existsSync(this.keyFile)) {
      throw new Error("Crypto not initialized. Run 'configsync init' first.");
    }

    const salt = fs.readFileSync(this.saltFile);
    const key = this.deriveKey(password, salt);

    const keyData = JSON.parse(fs.readFileSync(this.keyFile, 'utf-8'));
    const verificationBuf = Buffer.from(keyData.verification, 'base64');

    try {
      const decrypted = this.decryptWithKey(verificationBuf, key);
      if (decrypted.toString('utf-8') !== VERIFICATION_STRING) {
        throw new Error('Invalid password');
      }
    } catch {
      throw new Error('Invalid password');
    }

    this.derivedKey = key;
  }

  /**
   * Returns true when the manager has not yet been unlocked.
   */
  isLocked(): boolean {
    return this.derivedKey === null;
  }

  // ---------------------------------------------------------------------------
  // Encrypt / Decrypt (raw buffers)
  // ---------------------------------------------------------------------------

  /**
   * Encrypt arbitrary data with AES-256-GCM.
   * Returns: [IV (12)][authTag (16)][ciphertext]
   */
  encrypt(data: Buffer): Buffer {
    this.requireUnlocked();
    return this.encryptWithKey(data, this.derivedKey!);
  }

  /**
   * Decrypt data previously encrypted by `encrypt`.
   */
  decrypt(data: Buffer): Buffer {
    this.requireUnlocked();
    return this.decryptWithKey(data, this.derivedKey!);
  }

  // ---------------------------------------------------------------------------
  // Secret-level encrypt / decrypt (key-salted, base64 string I/O)
  // ---------------------------------------------------------------------------

  /**
   * Encrypt a secret value with an additional key-specific salt.
   * Returns a base64-encoded string.
   */
  encryptSecret(key: string, value: string): string {
    this.requireUnlocked();

    const keySalt = crypto.createHash('sha256').update(key).digest().subarray(0, 16);
    const saltedValue = Buffer.concat([keySalt, Buffer.from(value, 'utf-8')]);

    const encrypted = this.encrypt(saltedValue);
    return encrypted.toString('base64');
  }

  /**
   * Decrypt a secret value and verify its key-specific salt.
   */
  decryptSecret(key: string, encrypted: string): string {
    this.requireUnlocked();

    const encryptedBuf = Buffer.from(encrypted, 'base64');
    const decrypted = this.decrypt(encryptedBuf);

    const keySalt = crypto.createHash('sha256').update(key).digest().subarray(0, 16);
    if (!decrypted.subarray(0, 16).equals(keySalt)) {
      throw new Error('Invalid key for this secret');
    }

    return decrypted.subarray(16).toString('utf-8');
  }

  // ---------------------------------------------------------------------------
  // Password rotation
  // ---------------------------------------------------------------------------

  /**
   * Change the master password. Re-encrypts the secrets file with the new key.
   */
  changePassword(oldPassword: string, newPassword: string): void {
    // Ensure the old password is valid
    this.unlock(oldPassword);

    // Derive a new key with a fresh salt
    const newSalt = crypto.randomBytes(SALT_LENGTH);
    const newKey = this.deriveKey(newPassword, newSalt);

    // Re-encrypt the secrets file if it exists
    const secretsFile = path.join(this.configDir, 'secrets.enc');
    if (fs.existsSync(secretsFile)) {
      const raw = fs.readFileSync(secretsFile);
      const plaintext = this.decrypt(raw);
      const reEncrypted = this.encryptWithKey(plaintext, newKey);
      fs.writeFileSync(secretsFile, reEncrypted);
      fs.chmodSync(secretsFile, 0o600);
    }

    // Persist the new salt
    fs.writeFileSync(this.saltFile, newSalt);
    fs.chmodSync(this.saltFile, 0o600);

    // Store new verification token
    const verification = this.encryptWithKey(Buffer.from(VERIFICATION_STRING, 'utf-8'), newKey);
    const keyData = {
      verification: verification.toString('base64'),
      version: '2.0',
    };
    fs.writeFileSync(this.keyFile, JSON.stringify(keyData));
    fs.chmodSync(this.keyFile, 0o600);

    this.derivedKey = newKey;
  }

  // ---------------------------------------------------------------------------
  // Key derivation
  // ---------------------------------------------------------------------------

  /**
   * Derive an encryption key from a password and salt using PBKDF2.
   */
  deriveKey(password: string, salt: Buffer): Buffer {
    return crypto.pbkdf2Sync(password, salt, PBKDF2_ITERATIONS, KEY_LENGTH, PBKDF2_DIGEST);
  }

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  private requireUnlocked(): void {
    if (this.derivedKey === null) {
      throw new Error('CryptoManager is locked. Call unlock() first.');
    }
  }

  private encryptWithKey(data: Buffer, key: Buffer): Buffer {
    const iv = crypto.randomBytes(IV_LENGTH);
    const cipher = crypto.createCipheriv(ALGORITHM, key, iv);

    const encrypted = Buffer.concat([cipher.update(data), cipher.final()]);
    const authTag = cipher.getAuthTag();

    // Wire format: [iv][authTag][ciphertext]
    return Buffer.concat([iv, authTag, encrypted]);
  }

  private decryptWithKey(data: Buffer, key: Buffer): Buffer {
    const iv = data.subarray(0, IV_LENGTH);
    const authTag = data.subarray(IV_LENGTH, IV_LENGTH + AUTH_TAG_LENGTH);
    const ciphertext = data.subarray(IV_LENGTH + AUTH_TAG_LENGTH);

    const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
    decipher.setAuthTag(authTag);

    return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  }
}

File: devsync/crypto.py
"""Encryption and security management for DevSync"""

import base64
import json
import os
from pathlib import Path
from typing import Optional
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.backends import default_backend
import secrets
import hashlib


class CryptoManager:
    """Handles all encryption/decryption operations"""

    def __init__(self, config_dir: Path):
        self.config_dir = config_dir
        self.key_file = config_dir / ".key"
        self.salt_file = config_dir / ".salt"
        self._fernet: Optional[Fernet] = None

    def initialize(self, password: Optional[str] = None):
        """Initialize the crypto system with a master password"""
        if self.key_file.exists():
            raise ValueError("Crypto already initialized. Use 'unlock' instead.")

        if password is None:
            import getpass
            password = getpass.getpass("Enter master password: ")
            confirm = getpass.getpass("Confirm master password: ")
            if password != confirm:
                raise ValueError("Passwords don't match")

        # Generate a random salt
        salt = secrets.token_bytes(32)
        with open(self.salt_file, 'wb') as f:
            f.write(salt)

        # Derive key from password
        key = self._derive_key(password, salt)

        # Save key (encrypted with itself for verification)
        fernet = Fernet(key)
        verification = fernet.encrypt(b"DEVSYNC_VERIFICATION")

        key_data = {
            "verification": base64.b64encode(verification).decode(),
            "version": "1.0"
        }

        with open(self.key_file, 'w') as f:
            json.dump(key_data, f)

        self._fernet = fernet
        return True

    def unlock(self, password: Optional[str] = None) -> bool:
        """Unlock the crypto system with the master password"""
        if self._fernet is not None:
            return True  # Already unlocked

        if not self.key_file.exists():
            raise ValueError("Crypto not initialized. Run 'devsync init' first.")

        if password is None:
            import getpass
            password = getpass.getpass("Enter master password: ")

        # Load salt
        with open(self.salt_file, 'rb') as f:
            salt = f.read()

        # Derive key
        key = self._derive_key(password, salt)
        fernet = Fernet(key)

        # Verify password
        with open(self.key_file) as f:
            key_data = json.load(f)

        try:
            verification = base64.b64decode(key_data["verification"])
            decrypted = fernet.decrypt(verification)
            if decrypted != b"DEVSYNC_VERIFICATION":
                raise ValueError("Invalid password")
        except Exception:
            raise ValueError("Invalid password")

        self._fernet = fernet
        return True

    def is_locked(self) -> bool:
        """Check if crypto is locked"""
        return self._fernet is None

    def encrypt(self, data: bytes) -> bytes:
        """Encrypt data"""
        if self._fernet is None:
            self.unlock()
        return self._fernet.encrypt(data)

    def decrypt(self, data: bytes) -> bytes:
        """Decrypt data"""
        if self._fernet is None:
            self.unlock()
        return self._fernet.decrypt(data)

    def encrypt_secret(self, key: str, value: str) -> str:
        """Encrypt a secret value with additional key-specific encryption"""
        if self._fernet is None:
            self.unlock()

        # Add key-specific salt
        key_salt = hashlib.sha256(key.encode()).digest()[:16]
        salted_value = key_salt + value.encode()

        encrypted = self._fernet.encrypt(salted_value)
        return base64.b64encode(encrypted).decode()

    def decrypt_secret(self, key: str, encrypted: str) -> str:
        """Decrypt a secret value"""
        if self._fernet is None:
            self.unlock()

        encrypted_bytes = base64.b64decode(encrypted)
        decrypted = self._fernet.decrypt(encrypted_bytes)

        # Remove key-specific salt
        key_salt = hashlib.sha256(key.encode()).digest()[:16]
        if decrypted[:16] != key_salt:
            raise ValueError("Invalid key for this secret")

        return decrypted[16:].decode()

    def change_password(self, old_password: str, new_password: str):
        """Change the master password"""
        # Verify old password
        self.unlock(old_password)

        # Generate new salt
        new_salt = secrets.token_bytes(32)

        # Derive new key
        new_key = self._derive_key(new_password, new_salt)
        new_fernet = Fernet(new_key)

        # Re-encrypt all secrets with new key
        secrets_file = self.config_dir / "secrets.enc"
        if secrets_file.exists():
            # Decrypt with old key
            with open(secrets_file, 'rb') as f:
                secrets_data = self.decrypt(f.read())

            # Encrypt with new key
            with open(secrets_file, 'wb') as f:
                f.write(new_fernet.encrypt(secrets_data))

        # Save new salt
        with open(self.salt_file, 'wb') as f:
            f.write(new_salt)

        # Save new verification
        verification = new_fernet.encrypt(b"DEVSYNC_VERIFICATION")
        key_data = {
            "verification": base64.b64encode(verification).decode(),
            "version": "1.0"
        }

        with open(self.key_file, 'w') as f:
            json.dump(key_data, f)

        self._fernet = new_fernet
        return True

    def _derive_key(self, password: str, salt: bytes) -> bytes:
        """Derive encryption key from password"""
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=salt,
            iterations=100000,
            backend=default_backend()
        )
        key = base64.urlsafe_b64encode(kdf.derive(password.encode()))
        return key

    def export_secrets(self, output_file: Path, password: Optional[str] = None):
        """Export secrets to an encrypted file with optional password"""
        if password is None:
            import getpass
            password = getpass.getpass("Enter export password: ")

        secrets_file = self.config_dir / "secrets.enc"
        if not secrets_file.exists():
            raise ValueError("No secrets to export")

        # Read and decrypt secrets
        with open(secrets_file, 'rb') as f:
            secrets_data = self.decrypt(f.read())

        # Create export encryption
        export_salt = secrets.token_bytes(32)
        export_key = self._derive_key(password, export_salt)
        export_fernet = Fernet(export_key)

        # Encrypt for export
        export_data = {
            "salt": base64.b64encode(export_salt).decode(),
            "data": base64.b64encode(export_fernet.encrypt(secrets_data)).decode(),
            "version": "1.0"
        }

        with open(output_file, 'w') as f:
            json.dump(export_data, f, indent=2)

        return True

    def import_secrets(self, input_file: Path, password: Optional[str] = None):
        """Import secrets from an exported file"""
        if password is None:
            import getpass
            password = getpass.getpass("Enter import password: ")

        with open(input_file) as f:
            export_data = json.load(f)

        # Decrypt export
        export_salt = base64.b64decode(export_data["salt"])
        export_key = self._derive_key(password, export_salt)
        export_fernet = Fernet(export_key)

        encrypted_data = base64.b64decode(export_data["data"])
        secrets_data = export_fernet.decrypt(encrypted_data)

        # Re-encrypt with our key
        if self._fernet is None:
            self.unlock()

        secrets_file = self.config_dir / "secrets.enc"
        with open(secrets_file, 'wb') as f:
            f.write(self.encrypt(secrets_data))

        return True